File: usage-tracker.js
// ===========================================================================
// Usage Tracker - Moduł monitorowania użycia aplikacji
// ===========================================================================

const fs = require('fs');
const fsPromises = require('fs').promises;
const path = require('path');
const { app } = require('electron');

class UsageTracker {
    constructor() {
        this.dataPath = path.join(app.getPath('userData'), 'usage-stats.json');
        this.sessionStart = Date.now();
        this.data = this.loadSync(); // Sync tylko przy starcie
        this.saveQueue = Promise.resolve(); // Kolejka zapisu
    }

    // Załaduj dane synchronicznie (tylko przy starcie)
    loadSync() {
        try {
            if (fs.existsSync(this.dataPath)) {
                const raw = fs.readFileSync(this.dataPath, 'utf8');
                return JSON.parse(raw);
            }
        } catch (e) {
            console.log('Nie można załadować statystyk użycia:', e.message);
        }

        // Domyślne dane
        return {
            firstLaunch: new Date().toISOString(),
            launchCount: 0,
            totalSessionMinutes: 0,
            lastLaunch: null,
            loginEvents: []
        };
    }

    // Zapisz dane asynchronicznie (nie blokuje głównego wątku)
    async save() {
        // Używamy kolejki aby uniknąć równoczesnych zapisów
        this.saveQueue = this.saveQueue.then(async () => {
            try {
                await fsPromises.writeFile(
                    this.dataPath,
                    JSON.stringify(this.data, null, 2),
                    'utf8'
                );
            } catch (e) {
                console.log('Nie można zapisać statystyk użycia:', e.message);
            }
        });
        return this.saveQueue;
    }

    // Zarejestruj uruchomienie aplikacji
    trackLaunch() {
        this.data.launchCount++;
        this.data.lastLaunch = new Date().toISOString();
        this.sessionStart = Date.now();
        this.save();
    }

    // Zarejestruj zakończenie sesji
    trackSessionEnd() {
        const sessionMinutes = Math.round((Date.now() - this.sessionStart) / 60000);
        this.data.totalSessionMinutes += sessionMinutes;
        this.save();
    }

    // Zarejestruj wydarzenie logowania
    trackLogin(method) {
        this.data.loginEvents.push({
            date: new Date().toISOString(),
            method: method // 'code' lub 'phone_approval'
        });
        // Zachowaj tylko ostatnie 10 wydarzeń
        if (this.data.loginEvents.length > 10) {
            this.data.loginEvents = this.data.loginEvents.slice(-10);
        }
        this.save();
    }

    // Pobierz podsumowanie statystyk
    getSummary() {
        const hours = Math.floor(this.data.totalSessionMinutes / 60);
        const minutes = this.data.totalSessionMinutes % 60;

        return {
            firstLaunch: this.data.firstLaunch ? new Date(this.data.firstLaunch).toLocaleDateString('pl-PL') : 'Nieznana',
            lastLaunch: this.data.lastLaunch ? new Date(this.data.lastLaunch).toLocaleDateString('pl-PL') : 'Brak',
            launchCount: this.data.launchCount,
            totalTime: `${hours}h ${minutes}min`,
            loginCount: this.data.loginEvents.length,
            appVersion: app.getVersion() // Dynamicznie z package.json
        };
    }

    // Formatuj do wyświetlenia w dialogu
    getFormattedStats() {
        const s = this.getSummary();
        return `📊 Statystyki użycia J-Connect Enterprise

📅 Pierwsze uruchomienie: ${s.firstLaunch}
📅 Ostatnie uruchomienie: ${s.lastLaunch}

🚀 Liczba uruchomień: ${s.launchCount}
⏱️ Łączny czas użycia: ${s.totalTime}
🔐 Zarejestrowane logowania: ${s.loginCount}

📦 Wersja aplikacji: ${s.appVersion}`;
    }
}

module.exports = UsageTracker;

File: Services/UsageTracker.cs
using System.IO;
using System.Text.Json;

namespace AGMessenger.Services;

/// <summary>
/// Usage Tracker - tracks application usage statistics
/// Equivalent to usage-tracker.js from Electron version
/// </summary>
public class UsageTracker
{
    private static UsageTracker? _instance;
    private static readonly object _lock = new();

    public static UsageTracker Instance
    {
        get
        {
            lock (_lock)
            {
                _instance ??= new UsageTracker();
                return _instance;
            }
        }
    }

    private readonly string _dataPath;
    private UsageData _data;
    private DateTime _sessionStart;

    public int LaunchCount => _data.LaunchCount;

    private UsageTracker()
    {
        var appDataPath = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "AGMessenger");
        
        Directory.CreateDirectory(appDataPath);
        _dataPath = Path.Combine(appDataPath, "usage-stats.json");
        _sessionStart = DateTime.Now;
        _data = Load();
    }

    private UsageData Load()
    {
        try
        {
            if (File.Exists(_dataPath))
            {
                var json = File.ReadAllText(_dataPath);
                return JsonSerializer.Deserialize<UsageData>(json) ?? CreateDefaultData();
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Cannot load usage stats: {ex.Message}");
        }

        return CreateDefaultData();
    }

    private static UsageData CreateDefaultData()
    {
        return new UsageData
        {
            FirstLaunch = DateTime.Now,
            LaunchCount = 0,
            TotalSessionMinutes = 0,
            LastLaunch = null
        };
    }

    private void Save()
    {
        try
        {
            var json = JsonSerializer.Serialize(_data, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(_dataPath, json);
        }
        catch (Exception ex)
        {
            Console.WriteLine($"Cannot save usage stats: {ex.Message}");
        }
    }

    public void TrackLaunch()
    {
        _data.LaunchCount++;
        _data.LastLaunch = DateTime.Now;
        _sessionStart = DateTime.Now;
        Save();
    }

    public void TrackSessionEnd()
    {
        var sessionMinutes = (int)(DateTime.Now - _sessionStart).TotalMinutes;
        _data.TotalSessionMinutes += sessionMinutes;
        Save();
    }

    public string GetFormattedStats()
    {
        var hours = _data.TotalSessionMinutes / 60;
        var minutes = _data.TotalSessionMinutes % 60;
        var version = System.Reflection.Assembly.GetExecutingAssembly().GetName().Version;

        return $"""
            📊 Statystyki użycia FB-Messenger

            📅 Pierwsze uruchomienie: {_data.FirstLaunch:d}
            📅 Ostatnie uruchomienie: {_data.LastLaunch?.ToString("d") ?? "Brak"}

            🚀 Liczba uruchomień: {_data.LaunchCount}
            ⏱️ Łączny czas użycia: {hours}h {minutes}min

            📦 Wersja aplikacji: {version}
            """;
    }
}

public class UsageData
{
    public DateTime FirstLaunch { get; set; }
    public int LaunchCount { get; set; }
    public int TotalSessionMinutes { get; set; }
    public DateTime? LastLaunch { get; set; }
}
